app/openai_ops.py

```diff
@@ -359,76 +359,98 @@ def context_length(
         raise NotImplementedError(error)
 
 
-# Adapted from https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+def get_encoding_for_model(model: str):
+    try:
+        return tiktoken.encoding_for_model(model)
+    except KeyError:
+        return tiktoken.get_encoding("cl100k_base")
+
+
+def get_tokens_per_message(model: str):
+    model_specific_tokens = {
```

Comment:
this dict object can be global here to avoid generating the object every time this method is called